front/lib/api/oauth/providers/notion.ts

```diff
@@ -38,7 +42,145 @@ export class NotionOAuthProvider implements BaseOAuthStrategyProvider {
+      // For personal actions we fetch the workspace id of the admin-setup to enforce the workspace id to be the same as the admin-setup.
+      // workspace connection (setup by admin) if we have it.
+      const { mcp_server_id, ...restConfig } = extraConfig;
+
+      if (mcp_server_id) {
+        const mcpServerConnectionRes =
+          await MCPServerConnectionResource.findByMCPServer(auth, {
+            mcpServerId: mcp_server_id,
+            connectionType: "workspace",
+          });
+
+        if (mcpServerConnectionRes.isErr()) {
+          throw new Error(
+            "Failed to find MCP server connection: " +
+              mcpServerConnectionRes.error.message
+          );
+        }
+
+        const oauthApi = new OAuthAPI(config.getOAuthAPIConfig(), logger);
+        const connectionRes = await oauthApi.getConnectionMetadata({
+          connectionId: mcpServerConnectionRes.value.connectionId,
+        });
+        if (connectionRes.isErr()) {
+          logger.error(
+            "Failed to get connection metadata for admin-setup connection when updating the config for personal actions",
+            {
+              error: connectionRes.error,
+            }
+          );
+          throw new Error(
+            "Failed to get connection metadata: " + connectionRes.error.message
+          );
+        }
+
+        // The workspace_id in metadata is the Dust workspace ID, not the Notion workspace ID.
+        // We need to get the Notion workspace info from the raw OAuth response
+        const oauthRes = await oauthApi.getAccessToken({
+          connectionId: mcpServerConnectionRes.value.connectionId,
+        });
+        
+        if (oauthRes.isErr()) {
+          throw new Error(
+            "Failed to get access token for admin connection: " + oauthRes.error.message
+          );
+        }
+
+        // Extract Notion workspace info from the raw OAuth response
+        const notionWorkspaceId = (oauthRes.value.scrubbed_raw_json as any)?.workspace_id;
+        const notionWorkspaceName = (oauthRes.value.scrubbed_raw_json as any)?.workspace_name;
+
+        if (!notionWorkspaceId) {
+          throw new Error("No workspace_id found in admin OAuth response");
+        }
+
+        const updatedConfig = {
+          ...restConfig,
+          requested_notion_workspace_id: notionWorkspaceId,
+          requested_notion_workspace_name: notionWorkspaceName || "Unknown Workspace",
+        };
+
+        return updatedConfig;
+      }
+    }
+
+    return extraConfig;
+  }
+
+  async checkConnectionValidPostFinalize(connection: OAuthConnectionType) {
+    // If a Notion workspace was requested, we need to check that the Notion workspace id is the same as the requested one.
+    if ("requested_notion_workspace_id" in connection.metadata) {
+      const requestedNotionWorkspaceId = connection.metadata.requested_notion_workspace_id;
+      const requestedNotionWorkspaceName = connection.metadata.requested_notion_workspace_name;
```

Comment:
Here could we simplify to compare `requested_notion_workspace_id` with `workspace_id`? I don't have much context on this, so I looked how it's handled for Slack where it seems more simple. Is it relevant?

Maybe @Fraggle will know better 🚀

Reply:
Then LGTM (it's already approved)